NorthstarDLL/loghooks.cpp

```diff
@@ -0,0 +1,262 @@
+	HUD_PRINTCONSOLE,
+	HUD_PRINTTALK,
+	HUD_PRINTCENTER
+};
+
+class ICenterPrint
+{
+  public:
+	virtual void ctor() = 0;
+	virtual void Clear(void) = 0;
+	virtual void ColorPrint(int r, int g, int b, int a, wchar_t* text) = 0;
+	virtual void ColorPrint(int r, int g, int b, int a, char* text) = 0;
+	virtual void Print(wchar_t* text) = 0;
+	virtual void Print(char* text) = 0;
+	virtual void SetTextColor(int r, int g, int b, int a) = 0;
+};
+
+enum class SpewType_t
+{
+	SPEW_MESSAGE = 0,
+
+	SPEW_WARNING,
+	SPEW_ASSERT,
+	SPEW_ERROR,
+	SPEW_LOG,
+
+	SPEW_TYPE_COUNT
+};
+
+const std::unordered_map<SpewType_t, const char*> PrintSpewTypes = {
+	{SpewType_t::SPEW_MESSAGE, "SPEW_MESSAGE"},
+	{SpewType_t::SPEW_WARNING, "SPEW_WARNING"},
+	{SpewType_t::SPEW_ASSERT, "SPEW_ASSERT"},
+	{SpewType_t::SPEW_ERROR, "SPEW_ERROR"},
+	{SpewType_t::SPEW_LOG, "SPEW_LOG"}};
+
+// these are used to define the base text colour for these things
+const std::unordered_map<SpewType_t, spdlog::level::level_enum> PrintSpewLevels = {
+	{SpewType_t::SPEW_MESSAGE, spdlog::level::level_enum::info},
+	{SpewType_t::SPEW_WARNING, spdlog::level::level_enum::warn},
+	{SpewType_t::SPEW_ASSERT, spdlog::level::level_enum::err},
+	{SpewType_t::SPEW_ERROR, spdlog::level::level_enum::err},
+	{SpewType_t::SPEW_LOG, spdlog::level::level_enum::info}};
+
+const std::unordered_map<SpewType_t, const char> PrintSpewTypes_Short = {
+	{SpewType_t::SPEW_MESSAGE, 'M'},
+	{SpewType_t::SPEW_WARNING, 'W'},
+	{SpewType_t::SPEW_ASSERT, 'A'},
+	{SpewType_t::SPEW_ERROR, 'E'},
+	{SpewType_t::SPEW_LOG, 'L'}};
+
+ICenterPrint* pInternalCenterPrint = NULL;
+
+// clang-format off
+AUTOHOOK(TextMsg, client.dll + 0x198710,
+void,, (BFRead* msg))
+// clang-format on
+{
+	TextMsgPrintType_t msg_dest = (TextMsgPrintType_t)msg->ReadByte();
+
+	char text[256];
+	msg->ReadString(text, sizeof(text));
+
+	if (!Cvar_cl_showtextmsg->GetBool())
+		return;
+
+	switch (msg_dest)
+	{
+	case TextMsgPrintType_t::HUD_PRINTCENTER:
+		pInternalCenterPrint->Print(text);
+		break;
+
+	default:
+		spdlog::warn("Unimplemented TextMsg type {}! printing to console", msg_dest);
+		[[fallthrough]];
+
+	case TextMsgPrintType_t::HUD_PRINTCONSOLE:
+		auto endpos = strlen(text);
+		if (text[endpos - 1] == '\n')
+			text[endpos - 1] = '\0'; // cut off repeated newline
+
+		spdlog::info(text);
+		break;
+	}
+}
+
+// clang-format off
+AUTOHOOK(Hook_fprintf, engine.dll + 0x51B1F0,
+int,, (void* const stream, const char* const format, ...))
+// clang-format on
+{
+	va_list va;
+	va_start(va, format);
+
+	SQChar buf[1024];
+	int charsWritten = vsnprintf_s(buf, _TRUNCATE, format, va);
+
+	if (charsWritten > 0)
+	{
+		if (buf[charsWritten - 1] == '\n')
+			buf[charsWritten - 1] = '\0';
+		NS::log::NATIVE_EN->info("{}", buf);
+	}
+
+	va_end(va);
+	return 0;
+}
+
+// clang-format off
+AUTOHOOK(ConCommand_echo, engine.dll + 0x123680,
+void,, (const CCommand& arg))
+// clang-format on
+{
+	if (arg.ArgC() >= 2)
+		spdlog::info("[echo] {}", arg.ArgS());
```

Comment:
Should echo be using it's own logger? It has a colour defined, but it doesn't get used?

Reply:
Oh i kinda forgor about this tbf